indexer.go

```diff
@@ -0,0 +1,863 @@
+		return nil, err
+	}
+
+	// We're going to construct a fresh ProllyTree as our root to apply changes on
+	chunkConfig := tree.DefaultChunkConfig()
+	framework, err := tree.NewFramework(ctx, nodeStore, chunkConfig, nil)
+
+	if err != nil {
+		return nil, err
+	}
+
+	collections := map[string]*Collection{}
+
+	// TODO: Document this with an IPLD Schema
+	metadata, err := qp.BuildMap(basicnode.Prototype.Any, -1, func(am datamodel.MapAssembler) {
+		qp.MapEntry(am, VERSION_KEY, qp.Int(DB_VERSION))
+		qp.MapEntry(am, "format", qp.String("database"))
+	})
+
+	// Initialize the tree with metadata saying it's an indexed tryee
+	framework.Append(ctx, DB_METADATA_KEY, metadata)
+
+	tree, rootCid, err := framework.BuildTree(ctx)
+
+	if err != nil {
+		return nil, err
+	}
+
+	db := &Database{
+		&blockStore,
+		nodeStore,
+		rootCid,
+		tree,
+		collections,
+	}
+
+	return db, nil
+}
+
+func NewMemoryDatabase() (*Database, error) {
+	ctx := context.Background()
+	blockStore := blockstore.NewBlockstore(datastore.NewMapDatastore())
+
+	return NewDatabaseFromBlockStore(ctx, blockStore)
+}
+
+func (db *Database) Flush(ctx context.Context) error {
+	rootCid, err := db.tree.Rebuild(ctx)
+
+	if err != nil {
+		return err
+	}
+
+	db.rootCid = rootCid
+
+	return nil
+}
+
+func (db Database) StartMutating(ctx context.Context) error {
+	return db.tree.Mutate()
+}
+
+func (db Database) ExportToFile(ctx context.Context, destination string) error {
+	linkSystem := db.nodeStore.LinkSystem()
+	return car2.TraverseToFile(
+		ctx,
+		linkSystem,
+		db.rootCid,
+		selectorparse.CommonSelector_ExploreAllRecursively,
+		destination,
+	)
+}
+
+func (db *Database) Collection(name string, primaryKey ...string) (*Collection, error) {
+	if db.collections[name] == nil {
+		collection := Collection{
+			db,
+			name,
+			primaryKey,
+		}
+		db.collections[name] = &collection
+		collection.Initialize()
+	}
+
+	return db.collections[name], nil
+}
+
+func (collection Collection) HasPrimaryKey() bool {
+	return (collection.primaryKey != nil) && (len(collection.primaryKey) != 0)
+}
+
+func (collection Collection) Initialize() error {
+	// See if there is metadata (primary key, collection version number)
+	// If no metadata, take primary key and save it
+	return nil
+}
+
+func (collection Collection) IndexNDJSON(ctx context.Context, byteStream io.Reader) error {
+	err := collection.db.StartMutating(ctx)
+	if err != nil {
+		return err
+	}
+
+	scanner := bufio.NewScanner(byteStream)
+	// optionally, resize scanner's capacity for lines over 64K, see next example
+	for scanner.Scan() {
+		line := scanner.Text()
+		buffer := strings.NewReader(line)
+		mapBuilder := basicnode.Prototype.Map.NewBuilder()
+		err = dagjson.Decode(mapBuilder, buffer)
+
+		if err != nil {
+			return err
+		}
+
+		node := mapBuilder.Build()
+		err = collection.Insert(ctx, node)
+
+		if err != nil {
+			return err
+		}
+	}
+
+	err = scanner.Err()
+	// For some reason it throws an EOF even when closed properly
```

Comment:
This shouldn't be possible if you look at the scanner source code. I would look more into this to confirm it's occurring and why.